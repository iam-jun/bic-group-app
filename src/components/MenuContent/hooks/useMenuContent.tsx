/* eslint-disable @typescript-eslint/no-unused-vars */
import React from 'react';
import i18next from 'i18next';
import { isEmpty } from 'lodash';
import Clipboard from '@react-native-clipboard/clipboard';

import homeStack from '~/router/navigator/MainStack/stacks/homeStack/stack';
import { generateLink } from '~/utils/link';
import { IPost, PostType } from '~/interfaces/IPost';
import { IPayloadReactionDetailBottomSheet } from '~/interfaces/IModal';
import { Button } from '~/baseComponents';
import { useRootNavigation } from '../../../hooks/navigation';
import ReportContent from '~/components/Report/ReportContent';
import SeriesContentModal from '~/components/series/SeriesContentModal';
import useCommonController from '~/screens/store';
import { getRootGroupids } from '~/helpers/post';
import { ReportTo } from '~/interfaces/IReport';
import useModalStore from '~/store/modal';
import usePostsStore, { IPostsState } from '~/store/entities/posts';
import { onPressReportThisMember } from '~/helpers/blocking';
import quizStack from '~/router/navigator/MainStack/stacks/quizStack/stack';
import useQuizzesStore from '~/store/entities/quizzes';
import articleStack from '~/router/navigator/MainStack/stacks/articleStack/stack';
import seriesStack from '~/router/navigator/MainStack/stacks/series/stack';
import { getCopyLinkType, getReportContentType } from '../helper';
import useArticleController from '~/screens/articles/store';
import useNotificationItemMenu from '~/screens/Notification/components/NotificationMenu/store';
import { SpecificNotificationType } from '~/interfaces/INotification';

const useMenuContent = (
  data: IPost,
  contentType: PostType,
  isFromDetail: boolean,

  handleConfirmDeleteSeries?: () => void,
  handleDeletePostError?: (listAudiences: string[]) => void,
) => {
  const { rootNavigation } = useRootNavigation();

  const commonActions = useCommonController((state) => state.actions);
  const modalActions = useModalStore((state) => state.actions);
  const actionsQuizzesStore = useQuizzesStore((state) => state.actions);
  const articleControllerActions = useArticleController((state) => state.actions);
  const { deletePost } = usePostsStore((state: IPostsState) => state.actions);
  const specifictNotiActions = useNotificationItemMenu((state) => state.actions);

  const copyLinkType = getCopyLinkType(contentType);
  const reportContentType = getReportContentType(contentType);

  if (!data) return null;

  const {
    id: contentId,
    reactionsCount,
    audience,
    actor,
    quiz,
  } = data;

  const onPressEdit = () => {
    modalActions.hideModal();
    if (contentType === PostType.POST) {
      rootNavigation?.navigate?.(homeStack.createPost, {
        postId: contentId,
        replaceWithDetail: !isFromDetail,
      });
    }

    if (contentType === PostType.ARTICLE) {
      rootNavigation.navigate(articleStack.createArticle, { articleId: contentId });
    }

    if (contentType === PostType.SERIES) {
      rootNavigation?.navigate?.(
        seriesStack.createSeries, {
          seriesId: contentId,
          isFromDetail,
        },
      );
    }
  };

  const onPressEditSettings = () => {
    modalActions.hideModal();
    if (contentType === PostType.POST) {
      rootNavigation?.navigate?.(homeStack.postSettings, {
        postId: contentId,
        isFromPostMenuSettings: true,
      });
    }

    if (contentType === PostType.ARTICLE) {
      rootNavigation?.navigate?.(articleStack.createArticleSettings, {
        articleId: contentId,
        isFromArticleMenuSettings: true,
      });
    }

    if (contentType === PostType.SERIES) {
      rootNavigation?.navigate?.(seriesStack.seriesSettings, {
        seriesId: contentId,
        isFromSeriesMenuSettings: true,
      });
    }
  };

  const onPressSave = (isSaved: boolean) => {
    modalActions.hideModal();
    if (isSaved) {
      commonActions.unsavePost(contentId, contentType);
    } else {
      commonActions.savePost(contentId, contentType);
    }
  };

  const onPressCopyLink = () => {
    modalActions.hideModal();
    Clipboard.setString(generateLink(copyLinkType, contentId));
    modalActions.showToast({ content: 'common:text_link_copied_to_clipboard' });
  };

  const onPressViewReactions = () => {
    modalActions.hideModal();
    const firstReact = reactionsCount[0];
    if (!!firstReact && !isEmpty(firstReact)) {
      const initReaction = Object.keys(firstReact)[0];
      const payload: IPayloadReactionDetailBottomSheet = {
        reactionsCount,
        initReaction,
        getDataParam: { target: 'POST', targetId: contentId },
      };
      modalActions.showReactionDetailBottomSheet(payload);
    }
  };

  const onPressViewSeries = () => {
    modalActions.hideModal();

    modalActions.showModal({
      isOpen: true,
      isFullScreen: true,
      titleFullScreen: i18next.t('common:btn_view_series'),
      ContentComponent: <SeriesContentModal id={contentId} />,
    });
  };

  const onPressPin = () => {
    modalActions.hideModal();
    rootNavigation?.navigate?.(homeStack.pinContent, { postId: contentId });
  };

  const onPressReport = () => {
    const rootGroupIds = getRootGroupids(audience);

    modalActions.hideModal();

    // in this sprint default reportTo is COMMUNITY
    modalActions.showModal({
      isOpen: true,
      ContentComponent: (
        <ReportContent
          targetId={contentId}
          targetType={reportContentType}
          groupIds={rootGroupIds}
          reportTo={ReportTo.COMMUNITY}
        />
      ),
    });
  };

  const _onPressReportThisMember = () => {
    onPressReportThisMember({ modalActions, actor });
  };

  const onPressCUDQuiz = () => {
    modalActions.hideModal();
    rootNavigation?.navigate?.(quizStack.entryQuiz, { postId: contentId });
  };

  const onConfirmEditQuiz = () => {
    const onSuccess = () => {
      rootNavigation?.navigate(
        quizStack.composeQuiz, {
          quizId: quiz?.id,
        },
      );
    };
    actionsQuizzesStore.getQuizDetail({ quizId: quiz?.id, onSuccess });
  };

  const onPressEditQuiz = () => {
    modalActions.hideModal();

    modalActions.showAlert({
      title: i18next.t('quiz:alert_edit:header'),
      content: i18next.t('quiz:alert_edit:content'),
      cancelBtn: true,
      confirmLabel: i18next.t('quiz:continue'),
      onConfirm: onConfirmEditQuiz,
    });
  };

  const onPressDeleteQuiz = () => {
    modalActions.hideModal();

    modalActions.showAlert({
      title: i18next.t('quiz:alert_delete:header'),
      content: i18next.t('quiz:alert_delete:content'),
      cancelBtn: true,
      confirmLabel: i18next.t('common:btn_delete'),
      ConfirmBtnComponent: Button.Danger,
      confirmBtnProps: { type: 'ghost' },
      onConfirm: () => {
        actionsQuizzesStore.deleteQuiz(quiz?.id, contentId);
      },
    });
  };

  const onPressDeleteContent = () => {
    modalActions.hideModal();

    if (contentType === PostType.POST) {
      modalActions.showAlert({
        title: i18next.t('post:title_delete_post'),
        content: i18next.t('post:content_delete_post'),
        cancelBtn: true,
        confirmLabel: i18next.t('common:btn_delete'),
        ConfirmBtnComponent: Button.Danger,
        confirmBtnProps: { type: 'ghost' },
        onConfirm: () => deletePost({
          id: contentId,
          // callbackError: handleDeletePostError,
        }),
      });
    }

    if (contentType === PostType.ARTICLE) {
      modalActions.showAlert({
        title: i18next.t('article:menu:delete'),
        content: i18next.t('post:content_delete_article'),
        cancelBtn: true,
        confirmLabel: i18next.t('common:btn_delete'),
        ConfirmBtnComponent: Button.Danger,
        confirmBtnProps: { type: 'ghost' },
        onConfirm: () => articleControllerActions.deleteArticle(
          contentId,
        ),
      });
    }

    if (contentType === PostType.SERIES) {
      modalActions.showAlert({
        title: i18next.t('series:menu_text_delete_series'),
        content: i18next.t('series:content_delete_series'),
        cancelBtn: true,
        confirmLabel: i18next.t('common:btn_delete'),
        ConfirmBtnComponent: Button.Danger,
        confirmBtnProps: { type: 'ghost' },
        onConfirm: handleConfirmDeleteSeries,
      });
    }
  };

  const onPressNotificationSettingContent = (
    isEnable: boolean,
    contentTargetType: SpecificNotificationType,
  ) => {
    modalActions.hideModal();

    specifictNotiActions.editNotificationSettings(contentId, !isEnable, contentTargetType);
  };

  return {
    onPressEdit,
    onPressEditSettings,
    onPressSave,
    onPressCopyLink,
    onPressViewReactions,
    onPressViewSeries,
    onPressPin,
    onPressReport,
    _onPressReportThisMember,
    onPressCUDQuiz,
    onPressEditQuiz,
    onPressDeleteQuiz,
    onPressDeleteContent,
    onPressNotificationSettingContent,
  };
};

export default useMenuContent;
