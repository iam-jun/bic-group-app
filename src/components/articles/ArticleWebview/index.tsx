import { ExtendedTheme, useTheme } from '@react-navigation/native';
import React, {
  FC, useImperativeHandle, useRef,
} from 'react';
import {
  Platform, StyleProp, StyleSheet, View, ViewStyle,
} from 'react-native';
import WebView from 'react-native-webview';
import { CUSTOM_META, getInjectableJSMessage, USER_AGENT_DESKTOP } from '~/utils/link';
import { parseSafe } from '~/utils/common';
import useMounted from '~/hooks/mounted';
import getEnv from '~/utils/env';

export enum EventType {
    ON_LOAD_END = 'onLoadEnd', // must post content to editor after editor is mounted
    ON_INITIALIZE_END='onInitializeEnd', // after editor initiated content
    ON_SCROLL = 'onScroll'
}

const ARTICLE_EDITOR_URL = `https://${getEnv('SELF_DOMAIN')}/article/webview/v1`;
// const ARTICLE_EDITOR_URL = 'http://10.1.1.170:8088/article/webview';

export interface ArticleWebviewRef {
  injectJavaScript: (script: any) => void,
}

export interface ArticleWebviewProps {
  style?: StyleProp<ViewStyle>;
  webViewRef?: React.Ref<ArticleWebviewRef>;
  initScript: {type: string, payload: any};

  onScroll?: (event: {offsetY: number}) => void;
  onMessage: (message: {type: string, payload: any}) => void;
  onInitializeEnd?: () => void;
}

const ArticleWebview: FC<ArticleWebviewProps> = ({
  style,
  webViewRef,
  initScript,

  onScroll,
  onMessage,
  onInitializeEnd,
}: ArticleWebviewProps) => {
  const isMounted = useMounted();
  const theme: ExtendedTheme = useTheme();
  const styles = createStyle(theme);

  const _webViewRef = useRef<WebView>();

  // mention on android only work on desktop browser
  const userAgent = Platform.OS === 'android' ? USER_AGENT_DESKTOP : undefined;

  const injectJavaScript = (script: any) => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    _webViewRef?.current?.injectJavaScript?.(getInjectableJSMessage(script));
  };

  const blur = () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    _webViewRef?.current?.injectJavaScript?.('document.activeElement.blur()');
  };

  useImperativeHandle(
    webViewRef, () => ({
      injectJavaScript,
      blur,
    }),
  );

  const onLoadEnd = () => {
    injectJavaScript(initScript);
  };

  const _onInitializeEnd = () => {
    _webViewRef.current?.requestFocus();
    onInitializeEnd?.();
  };

  const _onMessage = (event: any) => {
    const message = parseSafe(event?.nativeEvent?.data);

    switch (message?.type) {
      case EventType.ON_LOAD_END:
        return onLoadEnd();
      case EventType.ON_INITIALIZE_END:
        return _onInitializeEnd?.();
      case EventType.ON_SCROLL:
        return onScroll?.(message?.payload);
      default:
        return onMessage(message);
    }
  };

  if (!isMounted) return null;

  return (
    <View testID="article_webview" style={[styles.container, style]}>
      <WebView
        testID="webview"
        key="article-webview"
        ref={_webViewRef}
        style={styles.webview}
        source={{ uri: ARTICLE_EDITOR_URL }}
        bounces
        cacheEnabled
        // avoid nested scroll
        overScrollMode="never"
        scrollEnabled={false}
        scalesPageToFit
        javaScriptEnabled
        domStorageEnabled
        startInLoadingState
        saveFormDataDisabled
        allowsFullscreenVideo
        hideKeyboardAccessoryView
        androidHardwareAccelerationDisabled
        nestedScrollEnabled={false}
        // force open native video player for the best performance
        allowsInlineMediaPlayback={false}
        userAgent={userAgent}
        /**
         * article detail may crash if androidLayerType set to hardware because of dynamic height
         * but article editor is very laggy if androidLayerType set to software
         */
        androidLayerType="hardware"
        injectedJavaScript={CUSTOM_META}
        showsVerticalScrollIndicator={false}
        showsHorizontalScrollIndicator={false}
        mediaPlaybackRequiresUserAction={false}
        automaticallyAdjustContentInsets={false}
        onMessage={_onMessage}
      />
    </View>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors } = theme;
  return StyleSheet.create({
    container: {
      flex: 1,
      backgroundColor: colors.white,
    },
    webview: {
      flex: 1,
      width: '100%',
      height: '100%',
      opacity: 0.99,
      overflow: 'hidden',
    },
  });
};

export default React.forwardRef((
  props: ArticleWebviewProps, ref: React.MutableRefObject<ArticleWebviewRef>,
) => (
  <ArticleWebview webViewRef={ref} {...props} />
));
