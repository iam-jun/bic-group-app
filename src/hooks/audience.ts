import {useSelector} from 'react-redux';
import {IObject} from '~/interfaces/common';

const useAudience = () => {
  const audiences = useSelector((state: IObject<any>) => state.audience);
  return audiences;
};

export default useAudience;
