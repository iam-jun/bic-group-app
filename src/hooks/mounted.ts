import { useEffect, useState } from 'react';
import { InteractionManager } from 'react-native';

const useMounted = (callback?: () => void) => {
  const [interactionsCompleted, setInteractionsCompleted] = useState(false);

  useEffect(() => {
    InteractionManager.runAfterInteractions(() => {
      setInteractionsCompleted(true);
      callback?.();
    });
  }, []);

  return interactionsCompleted;
};

export default useMounted;
