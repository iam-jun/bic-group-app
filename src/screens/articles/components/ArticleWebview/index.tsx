import { GiphyMedia } from '@giphy/react-native-sdk';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import React, {
  FC, useEffect, useRef, useState,
} from 'react';
import {
  Platform, StyleSheet, View,
} from 'react-native';
import { debounce } from 'lodash';
import WebView from 'react-native-webview';
import { CUSTOM_META, getInjectableJSMessage, USER_AGENT_DESKTOP } from '~/utils/link';
import MentionBar from '~/beinComponents/inputs/MentionInput/MentionBar';
import useMentionInputStore from '~/beinComponents/inputs/MentionInput/store';
import IMentionInputState from '~/beinComponents/inputs/MentionInput/store/Interface';
import KeyboardSpacer from '~/beinComponents/KeyboardSpacer';
import LoadingIndicator from '~/beinComponents/LoadingIndicator';
import StickerView from '~/components/StickerView';
import { IMentionUser } from '~/interfaces/IPost';
import { parseSafe } from '~/utils/common';
import getEnv from '~/utils/env';
import useMounted from '~/hooks/mounted';

enum EventType {
    ON_LOAD_END = 'onLoadEnd', // must post content to editor after editor is mounted
    ON_INITIALIZE_END='onInitializeEnd', // after editor initiated content return scrollHeight
    ON_EDITOR_CHANGE = 'onEditorChange',
    ON_SEARCH_MENTION = 'onSearchMention',
    ON_PRESS_BUTTON = 'onPressButton',
    ON_PRESS_MENTION = 'onPressMention'
}

const ARTICLE_EDITOR_URL = `https://${getEnv('SELF_DOMAIN')}/article/webview`;

export interface ArticleWebviewProps {
  articleData: any;
  readOnly?: boolean;

  onPressMentionAudience?: (user: IMentionUser) => void;
}

const ArticleWebview: FC<ArticleWebviewProps> = ({
  readOnly,
  articleData,

  onPressMentionAudience,
}: ArticleWebviewProps) => {
  const isMounted = useMounted();
  const theme: ExtendedTheme = useTheme();
  // const insets = useSafeAreaInsets();
  const styles = createStyle(theme);
  const webViewRef = useRef();
  const [isLoaded, setLoaded] = useState(false);
  const [webviewHeight, setWebviewHeight] = useState(0);
  const runSearch = useMentionInputStore((state: IMentionInputState) => state.doRunSearch);
  const setFullContent = useMentionInputStore((state: IMentionInputState) => state.setFullContent);

  const {
    content, groupIds, mentions, handleContentChange,
  } = articleData;

  const stickerViewRef = useRef<any>();

  const initScript = {
    type: 'initialize',
    payload: {
      readOnly,
      mentions,
      //   content: sampleData,
      content: parseSafe(content) || [],
    },
  };

  useEffect(() => {
    // reload webview after content change
    if (readOnly && isLoaded) {
      injectJavaScript(initScript);
    }
  }, [content]);

  const injectJavaScript = (script: any) => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    webViewRef?.current?.injectJavaScript?.(getInjectableJSMessage(script));
  };

  const onCompleteMention = (user: IMentionUser) => {
    injectJavaScript({
      type: 'insertMention',
      payload: {
        key: user.id,
        text: user.username,
        data: user,
      },
    });
  };

  const openGif = () => {
    stickerViewRef?.current?.show?.('giphy');
  };

  const onGifSelected = (gif: GiphyMedia) => {
    stickerViewRef?.current?.hide?.();

    injectJavaScript({
      type: 'insertNodes',
      payload: {
        type: 'img',
        children: [{ text: 'img' }],
        url: gif.url,
      },
    });
  };

  const onChangeContent = debounce((value) => {
    handleContentChange(value);
  }, 500);

  const onLoadEnd = () => {
    setLoaded(true);
    injectJavaScript(initScript);
  };

  const onInnitializeEnd = (payload: any) => {
    setWebviewHeight(payload?.scrollHeight);
  };

  const onChangeText = (payload: any) => {
    if (readOnly) return;
    setFullContent(payload?.rawContent);
    onChangeContent?.(JSON.stringify(payload?.contentState));
  };

  const onMention = (payload: any) => {
    if (readOnly) return;

    if (payload?.searchText) {
      runSearch(groupIds, payload?.searchText, true);
    }
  };

  const onPressButton = (type: string) => {
    if (type === 'gif') openGif();
  };

  const onMessage = (event: any) => {
    const message = parseSafe(event?.nativeEvent?.data);
    const payload = message?.payload;

    switch (message?.type) {
      case EventType.ON_LOAD_END:
        return onLoadEnd();
      case EventType.ON_INITIALIZE_END:
        return onInnitializeEnd(payload);
      case EventType.ON_EDITOR_CHANGE:
        return onChangeText(payload);
      case EventType.ON_SEARCH_MENTION:
        return onMention(payload);
      case EventType.ON_PRESS_BUTTON:
        return onPressButton(payload);
      case EventType.ON_PRESS_MENTION:
        return onPressMentionAudience?.(payload);
      default:
        return console.warn('Article webview onMessage unhandled', message);
    }
  };

  if (!isMounted) return <LoadingIndicator />;

  // mention on android only work on desktop browser
  const userAgent = Platform.OS === 'android' ? USER_AGENT_DESKTOP : undefined;

  return (
    <View style={styles.container}>
      { !isLoaded && <LoadingIndicator /> }
      <WebView
        ref={webViewRef}
        style={[styles.webview, readOnly && { height: webviewHeight }]}
        source={{ uri: ARTICLE_EDITOR_URL }}
          // source={{ uri: 'https://3e08-14-226-252-170.ap.ngrok.io/article/webview' }}
        useWebKit
        cacheEnabled
        scalesPageToFit
        javaScriptEnabled
        domStorageEnabled
        allowsFullscreenVideo
        saveFormDataDisabled
        hideKeyboardAccessoryView
        androidHardwareAccelerationDisabled
        scrollEnabled={!readOnly}
        nestedScrollEnabled={false}
        animationEnabled={false}
        // force open native video player for the best performance
        allowsInlineMediaPlayback={false}
        userAgent={userAgent}
        /**
         * article detail may crash if androidLayerType set to hardware because of dynamic height
         * but article editor is very laggy if androidLayerType set to software
         */
        androidLayerType={readOnly ? 'software' : 'hardware'}
        injectedJavaScript={CUSTOM_META}
        showsHorizontalScrollIndicator={false}
        mediaPlaybackRequiresUserAction={false}
        automaticallyAdjustContentInsets={false}
        onMessage={onMessage}
      />
      {!readOnly && (
        <View style={styles.mentions}>
          <MentionBar groupIds={groupIds} style={styles.mentionBar} onCompleteMention={onCompleteMention} />
          <StickerView
            stickerViewRef={stickerViewRef}
            onGifSelected={onGifSelected}
          />
          <KeyboardSpacer iosOnly />
        </View>
      )}
    </View>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors } = theme;
  return StyleSheet.create({
    container: {
      flex: 1,
      backgroundColor: colors.white,
    },
    scrollViewContainer: {
      flexGrow: 1,
    },
    webview: {
      width: '100%',
      height: '100%',
      opacity: 0.99,
      overflow: 'hidden',
    },
    mentions: {
      width: '100%',
      position: 'absolute',
      zIndex: 3,
      bottom: 0,
    },
    mentionBar: {
      height: 50,
    },
  });
};

export default ArticleWebview;
