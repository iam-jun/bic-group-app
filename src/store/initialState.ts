import {appInitState} from '~/store/app/reducer';
import {modalInitState} from '~/store/modal/reducer';
import {authInitState} from '~/screens/Auth/redux/reducer';
import {postInitState} from '~/screens/Post/redux/reducer';
import {groupInitState} from '~/screens/Groups/redux/reducer';
import {homeInitState} from '~/screens/Home/redux/reducer';
import {notiInitState} from '~/screens/Notification/redux/reducer';
import {menuInitState} from '~/screens/Menu/redux/reducer';
import {noInternetInitState} from '~/screens/NoInternet/redux/reducer';
import {mentionInputInitState} from '~/beinComponents/inputs/MentionInput/redux/reducer';

const initialState = {
  app: appInitState,
  modal: modalInitState,
  auth: authInitState,
  post: postInitState,
  groups: groupInitState,
  home: homeInitState,
  notifications: notiInitState,
  menu: menuInitState,
  noInternet: noInternetInitState,
  mentionInput: mentionInputInitState,
};

export default initialState;
