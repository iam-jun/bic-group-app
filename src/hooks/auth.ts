import {useDispatch, useSelector} from 'react-redux';
import {IObject} from '~/interfaces/common';
import {useEffect} from 'react';
import {Auth} from 'aws-amplify';
import authActions from '~/screens/Auth/redux/actions';
import {put} from 'redux-saga/effects';
import modalActions from '~/store/modal/actions';
import {useBaseHook} from '~/hooks/index';

const useAuth = () => {
  return useSelector((state: IObject<any>) => state.auth);
};

export const useUserIdAuth = () => {
  return (
    useSelector(
      (state: IObject<any>) =>
        state.auth?.user?.signInUserSession?.idToken?.payload?.[
          'custom:bein_user_id'
        ],
    ) || ''
  );
};

export const useAuthToken = () => {
  return (
    useSelector(
      (state: IObject<any>) =>
        state.auth?.user?.signInUserSession?.idToken?.jwtToken,
    ) || ''
  );
};

export const useAuthTokenExpire = () => {
  return (
    useSelector(
      (state: IObject<any>) =>
        state.auth?.user?.signInUserSession?.idToken?.payload?.exp,
    ) || ''
  );
};

export const useAuthKickOut = () => {
  const dispatch = useDispatch();
  const userId = useUserIdAuth();
  const {t} = useBaseHook();

  const checkAuthKickOut = async () => {
    try {
      await Auth.currentAuthenticatedUser({
        bypassCache: true,
      });
    } catch (e) {
      if (!userId) {
        return;
      }
      dispatch(authActions.signOut());
      dispatch(
        modalActions.showAlert({
          title: t('auth:text_kickout_title'),
          content: t('auth:text_kickout_desc'),
          onConfirm: () => put(modalActions.hideAlert()),
          confirmLabel: t('auth:text_kickout_confirm_button'),
        }),
      );
    }
  };

  useEffect(() => {
    checkAuthKickOut();
  }, []);
};

export default useAuth;
