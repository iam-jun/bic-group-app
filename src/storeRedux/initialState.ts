import { appInitState } from '~/storeRedux/app/reducer';
import { modalInitState } from '~/storeRedux/modal/reducer';
import { authInitState } from '~/storeRedux/auth/reducer';
import { postInitState } from '~/storeRedux/post/reducer';
import { groupInitState } from '~/storeRedux/groups/reducer';
import { homeInitState } from '~/storeRedux/home/reducer';
import { notiInitState } from '~/storeRedux/notification/reducer';
import { menuInitState } from '~/storeRedux/menu/reducer';
import { noInternetInitState } from '~/storeRedux/network/reducer';
import { mentionInputInitState } from '~/beinComponents/inputs/MentionInput/redux/reducer';
import { giphyInitState } from './giphy/reducer';

const initialState = {
  app: appInitState,
  modal: modalInitState,
  auth: authInitState,
  post: postInitState,
  groups: groupInitState,
  home: homeInitState,
  notifications: notiInitState,
  menu: menuInitState,
  noInternet: noInternetInitState,
  mentionInput: mentionInputInitState,
  giphy: giphyInitState,
};

export default initialState;
