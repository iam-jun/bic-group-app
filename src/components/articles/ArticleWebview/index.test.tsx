import React from 'react';

import {
  createTestStore,
  renderWithRedux,
} from '~/test/testUtils';
import ArticleWebview from '.';
import initialState from '~/storeRedux/initialState';
import * as useMounted from '~/hooks/mounted';

describe('ArticleWebview Component', () => {
  const state = { ...initialState };
  const onMessage = jest.fn();
  const initScript = { type: 'initDetail', payload: [] };

  it('renders correctly', () => {
    const store = createTestStore(state);

    const rendered = renderWithRedux(
      <ArticleWebview
        initScript={initScript}
        onMessage={onMessage}
      />, store,
    );
    expect(rendered).toMatchSnapshot();
  });

  it('should render null', () => {
    jest.spyOn(useMounted, 'default').mockImplementation(() => (false));

    const store = createTestStore({ ...initialState });
    const wrapper = renderWithRedux(
      <ArticleWebview
        initScript={initScript}
        onMessage={onMessage}
      />, store,
    );

    const webview = wrapper.queryByTestId('article_webview');

    expect(webview).toBeNull();
  });
});
