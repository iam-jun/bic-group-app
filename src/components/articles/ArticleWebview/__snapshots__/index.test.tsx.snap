// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleWebview Component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      },
      undefined,
    ]
  }
  testID="article_webview"
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <RNCWebView
      allowsFullscreenVideo={true}
      allowsInlineMediaPlayback={false}
      androidHardwareAccelerationDisabled={true}
      androidLayerType="hardware"
      automaticallyAdjustContentInsets={false}
      bounces={true}
      cacheEnabled={true}
      domStorageEnabled={true}
      hideKeyboardAccessoryView={true}
      injectedJavaScript="const meta = document.createElement('meta'); meta.setAttribute('content', 'width=device-width, initial-scale=1, maximum-scale=0.99, user-scalable=0'); meta.setAttribute('name', 'viewport'); document.getElementsByTagName('head')[0].appendChild(meta); "
      injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
      injectedJavaScriptForMainFrameOnly={true}
      javaScriptEnabled={true}
      mediaPlaybackRequiresUserAction={false}
      messagingEnabled={true}
      nestedScrollEnabled={false}
      onContentProcessDidTerminate={[Function]}
      onHttpError={[Function]}
      onLoadingError={[Function]}
      onLoadingFinish={[Function]}
      onLoadingProgress={[Function]}
      onLoadingStart={[Function]}
      onMessage={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      overScrollMode="never"
      saveFormDataDisabled={true}
      scalesPageToFit={true}
      scrollEnabled={false}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      source={
        Object {
          "uri": "https://undefined/article/webview/v1",
        }
      }
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#ffffff",
          },
          Object {
            "flex": 1,
            "height": "100%",
            "opacity": 0.99,
            "overflow": "hidden",
            "width": "100%",
          },
        ]
      }
      testID="webview"
      textInteractionEnabled={true}
      useSharedProcessPool={true}
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <ActivityIndicator />
    </View>
  </View>
</View>
`;
